### ISSUE TITLE
More simplification issues in tensorize 

### ISSUE THREAD

**Issue by derisavi-huawei** [dmlc/tvm#696]
Consider the following example. Basically, we are defining an algorithm and an intrinsic that are identical and we are trying to tensorize the outermost loop of the algorithm. 
```python
import tvm

def op(x):
    bh = 9
    bw = 9
    y = tvm.compute((bh, bw), lambda i,j: x[j/3 + i%3, j%3 + i/3])
    return y

def op_intrin(x):
    bh = 9
    bw = 9
    y = tvm.compute((bh, bw), lambda i,j: x[j/3 + i%3, j%3+ i/3])

    def ifunc(ins, outs):
        xx, = ins
        zz = outs[0]
        return tvm.call_packed("op", xx, zz)

    with tvm.build_config(offset_factor=2):
        return tvm.decl_tensor_intrin(y.op, ifunc)

A = tvm.placeholder((5, 5), name='A')
B = op(A)
bt = op_intrin(A)
s = tvm.create_schedule(B.op)
print(tvm.lower(s, [A, B], simple_mode=True))

x,y = B.op.axis
s[B].tensorize(x, bt)
s = s.normalize()
print(tvm.lower(s, [A, B], simple_mode=True))
```

TVM fails with the following error:
```
dmlc-core/include/dmlc/logging.h:308: [16:52:02] src/op/tensorize.cc:315: Check failed: Equal(lhs, rhs) Failed to match the compute with TensorIntrin tensor_intrin's declaration
  provided= A((((i % 3) + (j/3)) - (0/3)), (((i/3) + (j % 3)) - (0/3))), intrin=  A(((j/3) + (i % 3)), ((i/3) + (j % 3)))
```
There are two problems here: 
1) `0/3` is not simplified to zero. I can think of two solutions:
  a) Call Simplify before CanonicalSimplify in VerifyTensorizeBody() method in tensorize.cc. For this specific example, this solution works. However, if either of the expressions in the error message contain an op that exists in TVM only (i.e., and not in HalideIR) such as `reduce`, then Simplify call gives an error message. I think it would be nice to fix this (i.e., be able to call Simplify on any expression and just pass through non-HalideIR ops) anyway. What's the best way to fix this?
  b) Modify Canonical::Internal class to simplify Div (like what's done for Mod in #676 ). 

2) Sometimes `i%3 + j/3` is canonically-simplified to `i%3 + j/3` and sometimes to `j/3 + i%3`. The problem is in ComExprEntry::operator< that tries to order two terms with their pointers when everything else fails. Comparing pointers yield non-deterministic results. I can think of two solutions:
  a) before pointer of the two `ComExprEntry`s are compared, we can compare their `type_index()`. This is a partial solution because it will solve the problem for some cases (including the above) but not all. For example, canonicalizing `i%3 + j%3` will still be based on pointer values. I have this patch. 
  b) If all non-pointer comparisons cannot order the `ComExprEntry`s, compare them based on their string representations. For example, `i%4` will come before `j%3` and after `i%3` in the order.
For problem 2, I prefer implementing both solutions. Solution a is a fast check, and solution b is slow but handles all cases.

@tqchen, let me know please what you think and I'll send patches and testcases.

**Comment by tqchen** [comment]
@derisavi-huawei sorry for my delayed action on this because of nips, it seems that the recent proposed pr fixed this issue? If so, please add a regression test case and let us close this issue

**Comment by tqchen** [comment]
https://github.com/dmlc/tvm/pull/701

**Comment by derisavi-huawei** [comment]
@tqchen, #701 implements (2a) mentioned above. I also just submitted #704 to implement (2b). 
Issue (1) is still outstanding. I need your opinion on it. Thanks.

**Comment by tqchen** [comment]
For 1, it seems that we can simply jump in to support proper constant folder for CanonicalSimplify.

I will update comment on #704

**Comment by derisavi-huawei** [comment]
I understand you are suggesting to implement a stand-alone constant folder for CanonicalSimplify? 
- If yes, can't we somehow reuse HalideIR's simplifier? 
--  If yes, do you have a suggestion? I'm not sure what the best way is to do this.
--  If no, why?
- If no, can you elaborate?

**Comment by tqchen** [comment]
The general guideline when we designed CanonicalSimplify, is that it should be evolved to be self-sufficient for simplifying index expressions. So while it is possible to mix HalideIR's simplifier with this, we should patch it to have the same ability when it is easy to do so.  HalideIR's simplifier is complementary and can be used to simplify some more complicated expression including computation.

Eventually, we want to rethink the integer analysis here and build a unified simplifier that contains both symbolic integer and set analysis.

It is possible to also directly reuse HalideIR's simplifier. Note that so far Reduce is only used in very restricted case, as reduce is always the last expression of body. So we can simply have a function that strips reduce and use simplifier in the body

**Comment by derisavi-huawei** [comment]
I totally understand the need for a unified simplifier you're talking about. I can see that no matter how much we incrementally improve the current simplifier(s), there will still be fundamental limitations. However, implementing the unified simplifier needs a fairly significant time investment. So my attempts now are to solve short-term problems and get a few more things to work.
Following your last paragraph, I'm working on some changes so that we can call HalideIR's simplifier on "TVM expressions" basically by passing through any non-HalideIR ops and simplifying the rest using HalideIR's simplifer. Hopefully, this approach makes it future-proof as we may add non-HalideIR ops other than reduce in the future. I'll try to finish it asap and send it for your comments.

**Comment by tqchen** [comment]
in that case, add a callback of function that handles the default case should solve the problem.

**Comment by derisavi-huawei** [comment]
First attempt: I tried deriving from Halide::Internal::Simplify and adding `visit(const Reduce* op, const Expr& self)` method to it. The problem here is that `Reduce`'s `accept` gives a fatal error. (ir.cc) Why do we have two visitor styles? And why Reduce shouldn't work with the old style?

Second attempt: I tried to implement your idea above. I guess I need to use `IRMutator` and `mutate` instead of `IRVisitor` and `visit` since `visit` doesn't return a value as you show in your snippet. 
I am thinking of Halide::Internal::IRMutator::mutate like this:
```c++
Expr IRMutator::mutate(Expr e) {
    if (e.defined()) {
        Expr e2 = callback(this, e);
        if (e2.defined())
          expr = e2;
        else
          e.accept(this);
    } else {
        expr = Expr();
    }
    stmt = Stmt();
    return expr;
}
```
In which `callback` is a function that you have given an example of above, i.e., makes a `Reduce` node with `visit`ed children if it's a `Reduce` op at the top, otherwise return `Expr()`. I also need to get `callback` defined in IRMutator. Also we need a `callback` overloaded for `Stmt`. Is this what you had in mind?

**Comment by tqchen** [comment]
assume we can close this for now

**Comment by tqchen** [comment]
@derisavi-huawei can you submit a few regression tests on this?

**Comment by tqchen** [comment]
This is a good thread of issues, now things are fixed thanks to #720 #696 #718
